src/util/locktime.rs

```diff
@@ -0,0 +1,233 @@
+}
+
+impl LockTime {
+    /// Returns true if both lock times use the same unit i.e., both block based or both time based.
+    pub fn is_same_unit(&self, other: Self) -> bool {
+        core::mem::discriminant(self) == core::mem::discriminant(&other)
+    }
+
+    /// Returns true if this lock time value encodes a block height.
+    pub fn is_height(&self) -> bool {
+        match *self {
+            LockTime::Blocks(_) => true,
+            LockTime::Seconds(_) => false,
+        }
+    }
+
+    /// Returns true if this lock time value encodes a block time (UNIX timestamp).
+    pub fn is_time(&self) -> bool {
+        !self.is_height()
+    }
+
+    /// Returns the inner `u32` value.
+    ///
+    /// If you want to compare lock time values consider using `is_satisfied` instead of comparing
+    /// the results of `to_u32`.
+    pub fn to_u32(&self) -> u32 {
+        match *self {
+            LockTime::Blocks(ref t) => t.0,
+            LockTime::Seconds(ref t) => t.0,
+        }
+    }
+}
+
+impl From<u32> for LockTime {
+    /// Constructs a `LockTime` from an nLockTime value or the argument to OP_CHEKCLOCKTIMEVERIFY.
+    fn from(n: u32) -> Self {
+        if encodes_height(n) {
+            Self::Blocks(Height::new(n).expect("n is valid"))
+        } else {
+            Self::Seconds(Time::new(n).expect("n is valid"))
+        }
+    }
+}
+
+impl From<Height> for LockTime {
+    fn from(h: Height) -> Self {
+        LockTime::Blocks(h)
+    }
+}
+
+impl From<Time> for LockTime {
+    fn from(h: Time) -> Self {
+        LockTime::Seconds(h)
+    }
+}
+
+/// An absolute block height.
+#[derive(Debug, Clone, Copy, PartialEq, Eq, PartialOrd, Ord, Hash)]
+#[cfg_attr(feature = "serde", derive(Serialize, Deserialize))]
+#[cfg_attr(feature = "serde", serde(crate = "actual_serde"))]
+pub struct Height(u32);
+
+impl Height {
+    /// Constructs a new block height.
+    ///
+    /// # Errors
+    /// If `n` does not encode a block height value (see [`encodes_height`]).
+    pub fn new(n: u32) -> Result<Self, Error> {
+        if encodes_height(n) {
+            Ok(Self(n))
+        } else {
+            Err(Error::InvalidHeight)
+        }
+    }
+
+    /// Converts `Height` to its inner `u32` value.
+    pub fn to_u32(&self) -> u32 {
+        self.0
+    }
+}
+
+/// A UNIX timestamp, seconds since epoch.
+#[derive(Debug, Clone, Copy, PartialEq, Eq, PartialOrd, Ord, Hash)]
+#[cfg_attr(feature = "serde", derive(Serialize, Deserialize))]
+#[cfg_attr(feature = "serde", serde(crate = "actual_serde"))]
+pub struct Time(u32);
+
+impl Time {
+    /// Constructs a new block time.
+    ///
+    /// # Errors
+    /// If `n` does not encode a UNIX time stamp (see [`encodes_time`]).
+    pub fn new(n: u32) -> Result<Self, Error> {
+        if encodes_time(n) {
+            Ok(Self(n))
+        } else {
+            Err(Error::InvalidTime)
+        }
+    }
+
+    /// Converts `Time` to its inner `u32` value.
+    pub fn to_u32(&self) -> u32 {
```

Comment:
I used your suggestion but FTR I prefer the shorter version.

Reply:
Bother, this now makes no sense
```rust
    pub fn is_satisfied(&self, n_lock_time: LockTime) -> Result<bool, Error> {
        if !self.is_same_unit(n_lock_time) {
            return Err(Error::InvalidComparison);
        }
        Ok(n_lock_time.to_n_lock_time() >= self.to_n_lock_time())
    }
```

Reply:
```rust
pub fn is_ge_than(&self, other: LockTime) -> Result<bool, Error> {
    if !self.is_same_unit(lock_time) {
        return Err(Error::InvalidComparison);
   }
   Ok(other.to_consensus_u32() >= self.to_consensus_u32())
}
```

IMO `is_satisified` is a meh name. Confusing what "one LockTime satisifes another LockTime" means? then `lock_a.is_satisified(lock_b)` is not even clear which one is "satisfied".

After 5 minutes, I can't come up with a good name. Then I've noticed https://github.com/rust-bitcoin/rust-bitcoin/pull/994/files#r901348322 and I agree that I don't think it's just a good API.

First - the user call should probably never track time requirements in a `LockTime`, since it can only hold one unit at the time. I'd guess that most code will track it as a `(Option<Height>, Option<Time>)`.